state/indexer/sink/psql/psql.go

```diff
@@ -0,0 +1,253 @@
+// Package psql implements an event sink backed by a PostgreSQL database.
+package psql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/pubsub/query"
+	"github.com/tendermint/tendermint/types"
+)
+
+const (
+	tableBlocks     = "blocks"
+	tableTxResults  = "tx_results"
+	tableEvents     = "events"
+	tableAttributes = "attributes"
+	driverName      = "postgres"
+)
+
+// EventSink is an indexer backend providing the tx/block index services.  This
+// implementation stores records in a PostgreSQL database using the schema
+// defined in state/indexer/sink/psql/schema.sql.
+type EventSink struct {
+	store   *sql.DB
+	chainID string
+}
+
+// NewEventSink constructs an event sink associated with the PostgreSQL
+// database specified by connStr. Events written to the sink are attributed to
+// the specified chainID.
+func NewEventSink(connStr, chainID string) (*EventSink, error) {
```

Comment:
This is probably out of the scope of this CR, but it seems like any external caller of this would *likely* want to pass a connection object to share the connection pool, rather than forcing callers to have their own connection.

It also strikes me that we don't provide a way for callers to close this database connection/pool, which is maybe :/

Reply:
I thought about that too, and I generally agree. I mostly cargo-culted this interface from the existing psql implementation that I updated. I'd be happy to (separately) add a different constructor with a more rational interface. That should go first into tip, though, probably.

Reply:
yeah, totally agree, and think we can add this via a new change to the mainline branch.